File: src/tokenizer.py
from abc import ABC, abstractmethod
import tiktoken
class TokenCounter(ABC):
    @abstractmethod
    def count(self, text: str) -> int:
        pass
    @abstractmethod
    def encode(self, text: str) -> list[int]:
        pass
    @abstractmethod
    def decode(self, tokens: list[int]) -> str:
        pass
    

class TiktokenCounter(TokenCounter):
    def __init__(self, encoding_name: str = "cl100k_base"):
        self.encoder = tiktoken.get_encoding(encoding_name)
    def count(self, text: str) -> int:
        return len(self.encoder.encode(text))
    def encode(self, text: str) -> list[int]:
        return self.encoder.encode(text)
    def decode(self, token: list[int]) -> str:
        return self.encoder.decode(tokens)

counter = TiktokenCounter()
print(counter.count("It is recommended to install tiktoken within an activated virtual environment to prevent dependency conflicts:"))

File: src/tokenizer.ts
import { getEncoding } from "js-tiktoken";
import type { Tiktoken } from "js-tiktoken";

export interface TokenCounter {
  count(text: string): number;
  encode(text: string): number[];
  decode(tokens: number[]): string;
}

export class TiktokenCounter implements TokenCounter {
  private encoder: Tiktoken;

  constructor(encodingName: "cl100k_base" | "o200k_base" = "cl100k_base") {
    this.encoder = getEncoding(encodingName);
  }

  count(text: string): number {
    return this.encoder.encode(text).length;
  }

  encode(text: string): number[] {
    return this.encoder.encode(text);
  }

  decode(tokens: number[]): string {
    return this.encoder.decode(tokens);
  }
}
